mypy/messages.py

```diff
@@ -998,6 +999,184 @@ def untyped_decorated_function(self, typ: Type, context: Context) -> None:
     def typed_function_untyped_decorator(self, func_name: str, context: Context) -> None:
         self.fail('Untyped decorator makes function "{}" untyped'.format(func_name), context)
 
+    def bad_proto_variance(self, actual: int, tvar_name: str, expected: int,
+                           context: Context) -> None:
+        msg = capitalize("{} type variable '{}' used in protocol where"
+                         " {} one is expected".format(variance_string(actual),
+                                                      tvar_name,
+                                                      variance_string(expected)))
+        self.fail(msg, context)
+
+    def concrete_only_assign(self, typ: Type, context: Context) -> None:
+        self.fail("Can only assign concrete classes to a variable of type '{}'"
+                  .format(self.format(typ)), context)
+
+    def concrete_only_call(self, typ: Type, context: Context) -> None:
+        self.fail("Only concrete class can be given where '{}' is expected"
+                  .format(self.format(typ)), context)
+
+    def note_call(self, subtype: Type, call: Type, context: Context) -> None:
+        self.note("'{}.__call__' has type '{}'".format(strip_quotes(self.format(subtype)),
+                                                       self.format(call)), context)
+
+    def report_protocol_problems(self, subtype: Union[Instance, TupleType, TypedDictType],
+                                 supertype: Instance, context: Context) -> None:
+        """Report possible protocol conflicts between 'subtype' and 'supertype'.
+        This includes missing members, incompatible types, and incompatible
+        attribute flags, such as settable vs read-only or class variable vs
+        instance variable.
+        """
+        from mypy.subtypes import (get_missing_members, get_conflict_types, get_bad_flags,
+                                   is_subtype, IS_SETTABLE, IS_CLASSVAR, IS_CLASS_OR_STATIC)
+        OFFSET = 4  # Four spaces, so that notes will look like this:
+        # note: 'Cls' is missing following 'Proto' members:
+        # note:     method, attr
+        MAX_ITEMS = 2  # Maximum number of conflicts, missing members, and overloads shown
+        # List of special situations where we don't want to report additional problems
+        exclusions = {TypedDictType: ['typing.Mapping'],
+                      TupleType: ['typing.Iterable', 'typing.Sequence'],
+                      Instance: []}  # type: Dict[type, List[str]]
+        if supertype.type.fullname() in exclusions[type(subtype)]:
+            return
+
+        def pretty_overload(tp: Overloaded) -> None:
```

Comment:
I'd prefer this to be a normal method instead of a nested function. The enclosing method is pretty long. The same applies to `print_more`.